sns_driver.go

```diff
@@ -0,0 +1,93 @@
+package es
+
+import (
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+)
+
+// NewSNSDriver creates an SNSDriver
+func NewSNSDriver(client *sns.SNS, topicArn string, driver Driver) *SNSDriver {
+	return &SNSDriver{
+		client:   client,
+		topicArn: topicArn,
+		driver:   driver,
+	}
+}
+
+// SNSDriver creates a driver decorator that sends a notification to AWS' SNS
+// when new events are saved.
+type SNSDriver struct {
+	client   *sns.SNS
+	topicArn string
+	driver   Driver
+}
+
+type snsPacket struct {
+	Types []string
+}
+
+// Load delegates to internal driver
+func (d *SNSDriver) Load(aggregateID string) ([]*Event, error) {
+	return d.driver.Load(aggregateID)
+}
+
+// Save delegates to internal driver. If successful, it'll emit a single
+// notification with all event types.
+func (d *SNSDriver) Save(events []*Event) error {
+	err := d.driver.Save(events)
+	if err != nil {
+		return err
+	}
+	if len(events) == 0 {
+		return nil
+	}
+
+	types := d.extractEventTypes(events)
+	err = d.publish(snsPacket{Types: types})
+	if err != nil {
+		return err // TODO: return error? log? relaxed approach to notifying?
```

Comment:
Returning the error here would be a little misleading to the user of this driver… Because this is now a non-transactional take on notification, saving worked fine, but notifying failed. The idea is that this notification is not _so_ critical anymore… so perhaps logging `warn` could be enough?

Reply:
Yep, returning error here would also cascade into potential retries unnecessarily since the write operation into the source of truth has already succeeded. Log with warn level here sounds good.

Reply:
That's what I had in mind. 👍